File: 2117-find-original-array-from-doubled-array/find-original-array-from-doubled-array.cpp
class Solution {
public:
    vector<int> findOriginalArray(vector<int>& changed) {
        
        if( changed.size() % 2 != 0) return {};
        vector<int> original ;
        map<int,int> mp;
        sort(changed.begin(),changed.end());
        for(auto &it: changed)
        {
            mp[it]++;
        }

        for(auto &it: changed)
        {
            if( mp[it] == 0)
            {
                //doubled value may be 
                continue;
            }
            int doubled = it*2;
            if( mp.find(doubled) == mp.end() || mp[doubled] == 0 )
            {
                return {};
            }
            original.push_back(it);
            mp[it]--;
            mp[doubled]--;
        }
        return original;
    }
};

File: 2117-find-original-array-from-doubled-array/find-original-array-from-doubled-array.java
class Solution {
    public int[] findOriginalArray(int[] changed) {
        if( changed.length % 2 != 0) return new int[]{};
        int[] original = new int[changed.length/2];
        Map<Integer,Integer> mp = new HashMap<>();
        Arrays.sort(changed);
        for(int num: changed)
        {
            mp.put(num , mp.getOrDefault(num ,0 )+1);
        }
        int index = 0;
        for(int it: changed)
        {
            if( mp.get(it) == 0)
            {
                //doubled value may be 
                continue;
            }
            int doubled = it*2;
            if( !mp.containsKey(doubled) || mp.get(doubled)<=0 )
            {
                return new int[]{};
            }
            original[index++] = it;
            mp.put(it, mp.get(it)-1);
            mp.put(doubled,mp.get(doubled)-1);
        }
        return original;
    }
}